runtime/parachains/src/paras_inherent.rs

```diff
@@ -105,32 +181,154 @@ pub mod pallet {
 		const INHERENT_IDENTIFIER: InherentIdentifier = PARACHAINS_INHERENT_IDENTIFIER;
 
 		fn create_inherent(data: &InherentData) -> Option<Self::Call> {
-			let mut inherent_data: ParachainsInherentData<T::Header> =
-				match data.get_data(&Self::INHERENT_IDENTIFIER) {
-					Ok(Some(d)) => d,
-					Ok(None) => return None,
-					Err(_) => {
-						log::warn!(target: LOG_TARGET, "ParachainsInherentData failed to decode");
-
-						return None
-					},
-				};
+			let parent_hash = <frame_system::Pallet<T>>::parent_hash();
 
-			// filter out any unneeded dispute statements
-			T::DisputesHandler::filter_multi_dispute_data(&mut inherent_data.disputes);
+			let ParachainsInherentData::<T::Header> {
+				bitfields,
+				backed_candidates,
+				mut disputes,
+				parent_header,
+			} = match data.get_data(&Self::INHERENT_IDENTIFIER) {
+				Ok(Some(d)) => d,
+				Ok(None) => return None,
+				Err(_) => {
+					log::warn!(target: LOG_TARGET, "ParachainsInherentData failed to decode");
+					return None
+				},
+			};
 
-			// Sanity check: session changes can invalidate an inherent, and we _really_ don't want that to happen.
-			// See github.com/paritytech/polkadot/issues/1327
+			if parent_hash != parent_header.hash() {
+				log::warn!(
+					target: LOG_TARGET,
+					"ParachainsInherentData references a different parent header hash than frame"
+				);
+				return None
+			}
+
+			let current_session = <shared::Pallet<T>>::session_index();
+			let expected_bits = <scheduler::Pallet<T>>::availability_cores().len();
+			let validator_public = shared::Pallet::<T>::active_validator_keys();
+
+			T::DisputesHandler::filter_multi_dispute_data(&mut disputes);
+
+			let scheduled: Vec<CoreAssignment> = <scheduler::Pallet<T>>::scheduled();
```

Comment:
To recap what was discussed offline: reading the `scheduled` here will lead to unnecessarily filtering out valid backed candidates because the schedule is updated in `enter` with newly freed cores (from concluding disputes and fully available blocks). Instead we are going to re-create much of the logic from `enter` within the `with_transaction` closure and then read `scheduled`, guaranteeing that the block author won't have an early exit when sanitizing backed candidates.

Here is a test that I just added that currently fails but should work once the aforementioned changes are implemented: https://github.com/paritytech/polkadot/blob/zeke-create-enter-benchmarks-5/runtime/parachains/src/paras_inherent.rs#L845-L907 . (It uses the scenario builder from my benchmark branch so I didn't add it directly to this branch).